internal/goofys.go

```diff
@@ -1195,3 +1204,92 @@ func (fs *Goofys) Rename(
 	}
 	return
 }
+
+// return full name of the given inode
+func (fs *Goofys) GetFullName(id fuseops.InodeID) *string {
+	fs.mu.Lock()
+	inode := fs.inodes[id]
+	fs.mu.Unlock()
+	if inode == nil {
+		return nil
+	}
+	return inode.FullName()
+}
+
+func (fs *Goofys) GetFileSize(path string) (uint64, error) {
+	headOutput, err := fs.storageBackend.HeadBlob(&HeadBlobInput{Key: path})
+	if err != nil {
+		return 0, err
+	}
+	return headOutput.BlobItemOutput.Size, nil
+}
+
+func (fs *Goofys) DownloadToFile(ctx context.Context, key string, count uint64, target *os.File) (uint64, error) {
+	rawS3, err := fs.getRawS3()
+	if err == nil {
+		// No error means it's an S3 backend and we can use the downloader
+		s3Input := &s3.GetObjectInput{
+			Bucket: &fs.bucket,
+			Key:    &key,
+		}
+		if count > 0 {
+			s3Input.Range = aws.String(fmt.Sprintf("bytes=0-%v", count-1))
+		}
+		downloader := s3manager.NewDownloaderWithClient(rawS3)
+		written, err := downloader.DownloadWithContext(ctx, target, s3Input)
+		return uint64(written), err
+	}
+	// Otherwise just fall back to GetBlob()
+	getBlobOutput, err := fs.storageBackend.GetBlob(&GetBlobInput{
+		Key:   key,
+		Start: 0,
+		Count: count,
+	})
+	if err != nil {
+		return 0, err
+	}
+	written, err := io.Copy(target, getBlobOutput.Body)
+	getBlobOutput.Body.Close()
+	return uint64(written), err
+}
+
+// Expose the raw S3 object from the underlying 'storageBackend'.
+// Will error if the 'storageBackend' is not 'S3Backend' or 'GCS3' backend
+func (fs *Goofys) getRawS3() (*s3.S3, error) {
+	if s3Backend, ok := fs.storageBackend.Delegate().(*S3Backend); ok {
+		return s3Backend.S3, nil
+	}
+	if gcs3Backend, ok := fs.storageBackend.Delegate().(*GCS3); ok {
+		return gcs3Backend.S3, nil
+	}
+	return nil, fmt.Errorf("Backend Config is not S3")
+}
+
+// ListBlobs is a public method on goofys that lets callers iterate over files in the configured backend
+// and execute a function on each page of Blobs.
+// NOTE: The bulk of this was inspired and copied from internal/dir.go/listBlobsSafe()
+func (fs *Goofys) ListBlobs(ctx context.Context, prefix string, fn func([]BlobItemOutput) bool) error {
+	res := &ListBlobsOutput{
+		IsTruncated: true, // Just to get the loop started
+	}
+
+	for res.IsTruncated {
+		if ctx.Err() != nil {
+			break
+		}
+		listBlobsInput := &ListBlobsInput{
+			Prefix: &prefix,
+			// Get the continuation token from the prior result.
+			ContinuationToken: res.NextContinuationToken,
+		}
+		res, err := fs.storageBackend.ListBlobs(listBlobsInput)
```

Comment:
Has this been tested? My intuition is that this is a new `res` that will shadow the original and not actually affect the loop condition.

Reply:
fixed!